### ISSUE TITLE
Global Teardown

### ISSUE THREAD

**Issue by ChaseFlorell** [psake/psake#131]
I'm wondering about the need for a Global Teardown that executes after all tasks are complete? I wouldn't mind spiking it out, but I'm just wondering about everyone's opinion first.

Use Case:

> In our project, we compile each project individually (not the entire solution).   
> When the compilation happens, we capture all of the build warnings and dump them into a nicely formatted HTML table in an HTML file.  
> Each compilation get's it's own HTML file if there are build warnings, and these files are dropped into the Artifacts for TeamCity to pick up.
> 
> IF I had a Teardown function that ran at the end, I could get all of those html files, and format an index.html with links to the individual pages, and display that index page in a build warnings tab in TeamCity.

Essentially I see a need for something like

    Teardown {
        #do Stuff
    }

and then invoke the scriptblock at the end of the chain.

**Comment by vors** [comment]
I think you can achieve it without introducing new concept. You can create `task teamcityBuild` and add dependencies on tasks for all projects that need to be compile. Then inside the body you can `#do Stuff`. Perhaps I didn't understand your scenario correctly, please clarify if this approach doesn't solve the problem.

**Comment by ChaseFlorell** [comment]
For us, our builds are a little more complicated than normal (dependend builds and parallel builds and such), but I will give it some thought.

**Comment by ChaseFlorell** [comment]
*tl;dr;*
I will create a stand alone build step that runs at the end. That step will amalgamate the build warnings logs as needed
-----
So I've thought about this some more, and it won't work as you describe.

I've got the following tasks.

    Task default -depends BuildApp1, BuildApp2, BuildApp3
    Task BuildApp1 -depends CompileApp1, TestApp1, PackageApp1
    Task BuildApp2 -depends CompileApp2, TestApp2, PackageApp2
    Task BuildApp3 -depends CompileApp3, TestApp3, PackageApp3

Now on my local machine I'm able to run `> . \build.ps1`,  and get the default task to run... this takes (currently) about 15 minutes to run through all of the tests (mostly Specflow)

On the CI server (TeamCity), we wanted to parallelize the build, so we created dependent builds that run on separate build agents

    Task: BuildApp1        |        Posh command:  .\build.ps1 BuildApp1
    Task: BuildApp2        |        Posh command:  .\build.ps1 BuildApp2
    Task: BuildApp3        |        Posh command:  .\build.ps1 BuildApp3

All of these build tasks push the artifacts back to the parent build. What I'll do is have the parent build run a post build step to amalgamate the logs

    # only gets used on the developer workstation, TeamCity will execute each task independently
    Task default -depends BuildApp1, BuildApp2, BuildApp3, Teardown



    .\build.ps1 Teardown

**Comment by vors** [comment]
Do you have a code in parent build that waits for all child builds to be executed? If so, just add Teardown logic there. If not, then global Teardown would not help you, because tasks have been run on other machines, and psake doesn't target cross machine scenario.

**Comment by ChaseFlorell** [comment]
yeah, no I understand that. Like I said in my previous post, I think I have it sorted.
